lib/ansible/plugins/vars/host_group_vars.py

```diff
@@ -55,60 +55,92 @@
 
 import os
 from ansible.errors import AnsibleParserError
-from ansible.module_utils.common.text.converters import to_bytes, to_native, to_text
+from ansible.module_utils.common.text.converters import to_bytes, to_native
 from ansible.plugins.vars import BaseVarsPlugin
-from ansible.inventory.host import Host
-from ansible.inventory.group import Group
+from ansible.utils.path import basedir
 from ansible.utils.vars import combine_vars
 
+CANONICAL_PATHS = {}
 FOUND = {}  # type: dict[str, list[str]]
+NAK = set()
+HOSTS = set()
+GROUPS = set()
+
+
+def load_found_files(loader, data, found_files):
+    for found in found_files:
+        new_data = loader.load_from_file(found, cache=True, unsafe=True)
+        if new_data:  # ignore empty files
+            data = combine_vars(data, new_data)
+    return data
 
 
 class VarsModule(BaseVarsPlugin):
 
     REQUIRES_ENABLED = True
+    reuse_instance = True
 
     def get_vars(self, loader, path, entities, cache=True):
         ''' parses the inventory file '''
 
         if not isinstance(entities, list):
             entities = [entities]
 
-        super(VarsModule, self).get_vars(loader, path, entities)
+        # realpath is expensive
+        if path not in CANONICAL_PATHS:
+            CANONICAL_PATHS[path] = os.path.realpath(basedir(path))
+        realpath_basedir = CANONICAL_PATHS[path]
 
         data = {}
         for entity in entities:
-            if isinstance(entity, Host):
+            try:
+                if not entity.name or entity.base_type not in ('Host', 'Group'):
+                    raise AttributeError
+            except AttributeError:
+                raise AnsibleParserError("Supplied entity must be Host or Group, got %s instead" % (type(entity)))
```

Comment:
The first half of the conditional could falsely trip on an empty string or `None` `entity.name`. Since we're in a really hot loop, and this will end up being a fatal error anyway, it's usually better to code for the "happy path" of the loop without doing defensive checks upfront, and just wrap the places you want to provide better error handling with tightly-scoped exception handlers instead (as they're much cheaper at runtime).

Reply:
I fixed it to make the exception handling more specific. An empty name or None should be an error. When we checked the entity was a Group/Host using `isinstance`, empty/null/non-string names should be enforced at the inventory plugin level (assuming they are following the developer docs which will use [add_host,](https://github.com/ansible/ansible/blob/devel/lib/ansible/inventory/data.py#L191) and [add_group](https://github.com/ansible/ansible/blob/devel/lib/ansible/inventory/data.py#L160)).